sdks/java/harness/src/main/java/org/apache/beam/fn/harness/TransformProcessingThreadTracker.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * TransformProcessingThreadTracker tracks the thread ids for the transforms that are being
+ * processed in the SDK harness.
+ */
+public class TransformProcessingThreadTracker {
```

Comment:
Sorry that I'm not familiar with how logging service works, I'm wondering whether this will have multi-threading concurrency issue.

Reply:
I think there might be a very slight chance the processing thread moved onto another transform when the LogHandler haven't done transforming the log entries in previous one. But I think the it should be very rare(log transform should be almost instant) and I would argue that it's probably better to keep the logging just best practice instead of introducing locks to guarantee 100% metadata correctness?

Reply:
The argument for `it's probably better to keep the logging just best effort` is that it's ok to have step name mismatched with log message itself. Do you think it's acceptable when it happens?

Reply:
I'm not sure if it is ever gonna happen, we get to see if the integration test is flaky(in my 30+ IT test runs mismatch never happens). If the occurrence is less than 0.01% I don't think it'll have actual impact on usability. Current empty step in sdk logs have no values to users and can be considered almost 100% mismatch, so I think this PR should be at least an improvement to that.

Reply:
I agree that test signals can give us more confidence. 

> Current empty step in sdk logs have no values to users and can be considered almost 100% mismatch, so I think this PR should be at least an improvement to that.

I would say, `We provide some information but they can be wrong` is not better than `We don't provide more information`.

Reply:
I agree. I don't think the concurrency risk is ever gonna be an issue, we can tell from enabling the test and track the history(I also manually tested another 50 times and tests all passed with matching step). I believe `we provide some information and they can be wrong in very rare cases` would still be more valuable than don't provide the information and probably won't cause too much trouble for users, the element count in streaming pipeline falls into this best effort category as well.